eval.go

```diff
@@ -624,218 +324,402 @@ func newUserTypeInfo(userType reflect.Type) (*userTypeInfo, error) {
+				},
+			}, nil
+		}
+		return attrInfo{}, fmt.Errorf("user value field %s has unsupported type %s", field.Name, field.Type)
 	default:
 		return attrInfo{}, fmt.Errorf("user value field %s has unsupported type %s", field.Name, field.Type)
 	}
 }
 
-// attrInfo returns information on the attribute with the given name.
-func (tinfo *userTypeInfo) attrInfo(name string) attrInfo {
-	info, ok := tinfo.fields[name]
-	if ok {
-		return info
-	}
-	if tinfo.getAttribute == nil {
-		return attrInfo{
-			asString: func(v reflect.Value) string {
-				return ""
-			},
-			asBytes: func(v reflect.Value) []byte {
-				return nil
-			},
-		}
-	}
-	return attrInfo{
-		asString: func(v reflect.Value) string {
-			return tinfo.getAttribute(v, name)
-		},
-		asBytes: func(v reflect.Value) []byte {
-			return []byte(tinfo.getAttribute(v, name))
-		},
-		asSemver: func(v reflect.Value) (*semver.Version, error) {
-			return parseSemver(tinfo.getAttribute(v, name))
-		},
+func isPredefined(attr string) bool {
+	switch attr {
+	case "Identifier", "Email", "Country":
+		return true
 	}
+	return false
 }
 
-func parseSemver(s string) (*semver.Version, error) {
-	if s == "" {
-		return nil, nil
-	}
-	vers, err := semver.Parse(s)
-	if err != nil {
-		return nil, err
-	}
-	return &vers, nil
+func (t *userTypeInfo) getString(v reflect.Value, attr string) (string, bool, error) {
+	info, ok := t.fields[attr]
+	if ok && info.asString != nil {
+		result, converted := info.asString(v)
+		if len(result) == 0 && isPredefined(attr) {
+			return "", false, &userAttrMissingError{attr: attr}
+		}
+		return result, converted, nil
+	} else if t.getAttribute != nil {
+		res := t.getAttribute(v, attr)
+		if res == nil {
+			return "", false, &userAttrMissingError{attr: attr}
+		}
+		switch val := res.(type) {
+		case string:
+			return val, false, nil
+		case []byte:
+			return string(val), false, nil
+		case []string:
+			b, _ := toJson(val)
+			return string(b), false, nil
+		case float32:
+			return strconv.FormatFloat(float64(val), 'g', -1, 64), true, nil
+		case float64:
+			return strconv.FormatFloat(val, 'g', -1, 64), true, nil
+		case int:
+			return strconv.FormatInt(int64(val), 10), true, nil
+		case uint:
+			return strconv.FormatUint(uint64(val), 10), true, nil
+		case int8:
+			return strconv.FormatInt(int64(val), 10), true, nil
+		case uint8:
+			return strconv.FormatUint(uint64(val), 10), true, nil
+		case int16:
+			return strconv.FormatInt(int64(val), 10), true, nil
+		case uint16:
+			return strconv.FormatUint(uint64(val), 10), true, nil
+		case int32:
+			return strconv.FormatInt(int64(val), 10), true, nil
+		case uint32:
+			return strconv.FormatUint(uint64(val), 10), true, nil
+		case int64:
+			return strconv.FormatInt(val, 10), true, nil
+		case uint64:
+			return strconv.FormatUint(val, 10), true, nil
+		case uintptr:
+			return strconv.FormatUint(uint64(val), 10), true, nil
+		case time.Time:
+			return strconv.FormatInt(val.UnixMilli()/1000, 10), true, nil
```

Comment:
As per spec, datetime value is a

> unix timestamp in seconds (**with an optional decimal part**)

```suggestion
			return strconv.FormatFloat(float64(val.UnixMilli())/1000.0, 'f', -1, 64), true, nil
```